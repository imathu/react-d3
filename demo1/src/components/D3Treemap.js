import * as d3 from 'd3';

class D3Component {

  containerEl;
  props;
  svg;

  constructor(containerEl, props, title=false) {
    this.containerEl = containerEl;
    this.props = props;
    const { width, height, data } = props;
    this.data = data;
    this.title = title;
    this.svg = d3.select(containerEl)
      .append('svg')
      .style('background-color', 'white')
      .attr('width', width)
      .attr('height', height);
    this.updateDatapoints();
  }

  draw = (width, height) => {
    const { svg } = this;
    d3.treemap()
    .paddingTop((this.title) ? 28 : 0)
    .paddingRight(7)
    .paddingInner(3) 
    .size([width, height])(this.root)

    svg
    .selectAll("rect")
    .data(this.root.leaves())
    .enter()
    .append("rect")
      .attr('x', function (d) { return d.x0; })
      .attr('y', function (d) { return d.y0; })
      .attr('width', function (d) { return d.x1 - d.x0; })
      .attr('height', function (d) { return d.y1 - d.y0; })
      .style("stroke", "black")
      .style("fill", d => {
        return (d.data.state === true) ? "#43C275" : "#C5202A";
      })

    svg
      .selectAll("text")
      .data(this.root.leaves())
      .enter()
      .append("text")
        .attr("x", function(d){ return d.x0+5})    // +10 to adjust position (more right)
        .attr("y", function(d){ return d.y0+20})    // +20 to adjust position (lower)
        .text(function(d){ return d.data.name })
        .attr("font-size", "15px")
        .attr("fill", "white")

      if (this.title) {
        svg
      .selectAll("titles")
      .data(this.root.descendants().filter(function(d){return d.depth===1}))
      .enter()
      .append("text")
        .attr("x", function(d){ return d.x0})
        .attr("y", function(d){ return d.y0+21})
        .text(function(d){ return d.data.name })
        .attr("font-size", "19px")
        .attr("fill",  "gray")
      }
  }

  updateDatapoints = () => {
    const { props: { width, height, data } } = this;
    const root = d3.hierarchy(data).sum(function(d){ return d.value})
          this.root = root;
          this.draw(width, height);        
  }

  setActiveDatapoint = (d, node) => {
    d3.select(node).style('fill', 'yellow');
    this.props.onDatapointClick(d);
  }

  resize = (width, height) => {
    const { svg } = this;
    svg.attr('width', width)
    .attr('height', height);
    svg.selectAll("*").remove();
    this.draw(width, height);
  }
}

export default D3Component;