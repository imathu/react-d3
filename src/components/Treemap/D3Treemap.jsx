import React, { useState, useEffect } from 'react';
import { PropTypes } from 'prop-types';
import { Popup } from 'semantic-ui-react';
import uuid from 'uuid';
import * as d3 from 'd3';

// S3 treemap example for React
// using React for manipulating the DOM
// using S3 for calculating the treemap and layout

import './Treemap.css';

const Node = ({ d, color }) => (
  <Popup
    content={`nodename: ${d.data.name}`}
    position="right center"
    inverted
    on="hover"
    style={{ borderRadius: 3, opacity: 0.8, padding: '2' }}
    trigger={(
      <rect
        className="node"
        style={{
          x: d.x0,
          y: d.y0,
          width: d.x1 - d.x0,
          height: d.y1 - d.y0,
          stroke: color,
          fill: color,
        }}
      />
    )}
  />
);
Node.propTypes = {
  d: PropTypes.shape({
    data: PropTypes.shape({
      name: PropTypes.string.isRequired,
    }).isRequired,
    x0: PropTypes.number.isRequired,
    x1: PropTypes.number.isRequired,
    y0: PropTypes.number.isRequired,
    y1: PropTypes.number.isRequired,
  }).isRequired,
  color: PropTypes.string.isRequired,
};

const D3Treemap = ({
  data, width, height, title,
}) => {
  const [treemapData, setTreemapData] = useState(null);

  useEffect(() => {
    // initialize the treemap using s3 hierarchy & treemap functions
    const hierarchy = d3.hierarchy(data).sum((d) => d.value);
    const treemap = d3.treemap()
      .paddingTop((title) ? 28 : 1)
      .paddingBottom(1)
      .paddingRight(7)
      .paddingInner(4)
      .size([width, height])(hierarchy);
    setTreemapData(treemap);
  }, [data, width, height, title]);

  return (
    <svg style={{ backgroundColor: 'white', width, height }}>
      {treemapData && treemapData.leaves().map((d) => (
        <Node key={uuid.v4()} d={d} color={(d.data.state === true) ? '#D2222D' : '#238823'} />
      ))}
      {treemapData && treemapData.leaves().map((d) => (
        <text key={uuid.v4()} x={d.x0 + 5} y={d.y0 + 20} fontSize="15px" fill="lightgray">{d.data.name}</text>
      ))}
      {title && treemapData && treemapData.descendants().filter((d) => d.depth === 1).map((d) => (
        <text key={uuid.v4()} x={d.x0} y={d.y0 + 21} fontSize="19px" fill="gray">{d.data.name}</text>
      ))}
    </svg>
  );
};
D3Treemap.propTypes = {
  data: PropTypes.shape({}).isRequired,
  height: PropTypes.number.isRequired,
  width: PropTypes.number.isRequired,
  title: PropTypes.bool,
};
D3Treemap.defaultProps = {
  title: false,
};

export default D3Treemap;
