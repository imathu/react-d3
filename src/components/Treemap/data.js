export const data1 = {
  children: [{
    name: 'Test',
    children: [{
      name: 'a',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'b',
      group: 'A',
      value: 1,
      state: false,
    }, {
      name: 'c',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'd',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'e',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'f',
      group: 'F',
      value: 1,
      state: false,
    }, {
      name: 'g',
      group: 'G',
      value: 1,
      state: true,
    }, {
      name: 'h',
      group: 'H',
      value: 1,
      state: false,
    }, {
      name: 'i',
      group: 'I',
      value: 1,
      state: true,
    }, {
      name: 'j',
      group: 'J',
      value: 1,
      state: true,
    }, {
      name: 'a',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'b',
      group: 'A',
      value: 1,
      state: false,
    }, {
      name: 'c',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'd',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'e',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'f',
      group: 'F',
      value: 1,
      state: false,
    }, {
      name: 'g',
      group: 'G',
      value: 1,
      state: true,
    }, {
      name: 'h',
      group: 'H',
      value: 1,
      state: false,
    }, {
      name: 'i',
      group: 'I',
      value: 1,
      state: true,
    }, {
      name: 'j',
      group: 'J',
      value: 1,
      state: true,
    }, {
      name: 'a',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'b',
      group: 'A',
      value: 1,
      state: false,
    }, {
      name: 'c',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'd',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'e',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'f',
      group: 'F',
      value: 1,
      state: false,
    }, {
      name: 'g',
      group: 'G',
      value: 1,
      state: true,
    }, {
      name: 'h',
      group: 'H',
      value: 1,
      state: false,
    }, {
      name: 'i',
      group: 'I',
      value: 1,
      state: true,
    }, {
      name: 'j',
      group: 'J',
      value: 1,
      state: true,
    }],
  }],
};

export const data2 = {
  children: [{
    name: 'APP 0',
    children: [{
      name: 'a',
      group: 'A',
      value: 1,
      state: true,
    }, {
      name: 'b',
      group: 'B',
      value: 1,
      state: false,
    }],
  }, {
    name: 'APP 1',
    children: [{
      name: 'c1',
      group: 'APP 1',
      value: 2,
      state: true,
    }, {
      name: 'c2',
      group: 'APP 1',
      value: 3,
      state: false,
    }, {
      name: 'c3',
      group: 'APP 1',
      value: 1,
      state: true,
    }],
  }, {
    name: 'APP 2',
    children: [{
      name: 'd',
      group: 'D',
      value: 1,
      state: true,
    }, {
      name: 'e',
      group: 'E',
      value: 1,
      state: true,
    }, {
      name: 'f',
      group: 'F',
      value: 1,
      state: false,
    }, {
      name: 'g',
      group: 'G',
      value: 1,
      state: true,
    }, {
      name: 'h',
      group: 'H',
      value: 1,
      state: false,
    }, {
      name: 'i',
      group: 'I',
      value: 1,
      state: true,
    }, {
      name: 'j',
      group: 'J',
      value: 1,
      state: true,
    }, {
      name: 'e',
      group: 'E',
      value: 1,
      state: true,
    }, {
      name: 'f',
      group: 'F',
      value: 1,
      state: false,
    }, {
      name: 'g',
      group: 'G',
      value: 1,
      state: true,
    }, {
      name: 'h',
      group: 'H',
      value: 1,
      state: false,
    }, {
      name: 'i',
      group: 'I',
      value: 1,
      state: true,
    }, {
      name: 'j',
      group: 'J',
      value: 1,
      state: true,
    }],
  }],
  name: 'Test',
};
