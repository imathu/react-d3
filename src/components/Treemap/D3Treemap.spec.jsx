import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import D3Treemap from './D3Treemap';

import { data1 } from './data';

describe('D3Treemap', () => {
  test('snapshot renders', () => {
    const component = renderer.create(<D3Treemap data={data1} height={300} width={200} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('count nodes', () => {
    const wrapper = mount(<D3Treemap data={data1} height={300} width={200} />);
    expect(wrapper.find('.node').length).toBe(30);
  });
});
