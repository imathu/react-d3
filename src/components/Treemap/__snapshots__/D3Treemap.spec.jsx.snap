// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`D3Treemap snapshot renders 1`] = `
<svg
  style={
    Object {
      "backgroundColor": "white",
      "height": 300,
      "width": 200,
    }
  }
/>
`;
